### PR TITLE
fix: Should properly look for element when removing

### LINKED ISSUE #2973: Children are not ordered by priority following add and remove
# Current bug behavior
After a particular sequence of adding and removing component children with different priority values, the children will become ordered incorrectly resulting in incorrect rendering.

# Expected behavior
Children are always ordered and rendered by priority.

# Steps to reproduce
The following example reproduces the issue on flame `1.14.0`:

```dart
import 'package:flame/components.dart';
import 'package:flame/game.dart';
import 'package:flame/input.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(
    GameWidget(
      game: MyGame(),
    ),
  );
}

class MyGame extends FlameGame {
  @override
  Future<void> onLoad() async {
    camera.viewport.add(ButtonComponent(
      button: TextComponent(text: 'Go'),
      onPressed: () async {
        final redCircle = CircleComponent(
          position: Vector2(-512, 0),
          radius: 128,
          priority: 10,
          paint: Paint()..color = const Color(0xffff0000),
        );
        await world.add(redCircle);

        await Future.delayed(const Duration(seconds: 1));
        final greenCircle = CircleComponent(
          position: Vector2(0, 0),
          radius: 128,
          priority: 9,
          paint: Paint()..color = const Color(0xff00ff00),
        );
        await world.add(greenCircle);

        await Future.delayed(const Duration(seconds: 1));
        final yellowCircle = CircleComponent(
          position: Vector2(-512, 0),
          radius: 128,
          priority: 10,
          paint: Paint()..color = const Color(0xff0000ff),
        );
        world.remove(redCircle);
        await world.add(yellowCircle);

        await Future.delayed(const Duration(seconds: 1));
        final purpleCircle = CircleComponent(
          position: Vector2(64, 0),
          radius: 128,
          priority: 8,
          paint: Paint()..color = const Color(0xffff00ff),
        );
        await world.add(purpleCircle);

        await Future.delayed(const Duration(seconds: 1));
        print(world.children.map((p) => p.priority).toList());
        // Outputs:
        // [9, 8, 10]
      },
    ));
  }
}
```

After the sequence is done:
1. Children are no longer in priority order.
2. The purple circle (priority 8) is incorrectly rendered on top of the green circle (priority 9).
  Screenshot 2024-01-11 at 1 36 35 PM

Note if we then call `world.children.rebalanceAll()` the children become ordered correctly again.

# Flutter doctor output

```
[✓] Flutter (Channel stable, 3.13.9, on macOS 14.2.1 23C71 darwin-arm64, locale en-US)
    • Flutter version 3.13.9 on channel stable at /opt/homebrew/Caskroom/flutter/3.13.7/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision d211f42860 (3 months ago), 2023-10-25 13:42:25 -0700
    • Engine revision 0545f8705d
    • Dart version 3.1.5
    • DevTools version 2.25.0
```

### PR DESCRIPTION
**Pull request by spydon**
This fixes https://github.com/flame-engine/flame/issues/2973

### CODE REVIEW

**Review by erickzanardo — APPROVED** [review]

File: test/ordered_set_test.dart [hunk 1]
```diff
@@ -320,14 +320,24 @@ void main() {
 
       test('sorts after remove', () {
         final orderedSet = OrderedSet<int>();
-        // The initial elements must be in order.
+        // The initial elements must be in order to reproduce the issue.
```
Comment (luanpotter) [inline test/ordered_set_test.dart:323]: what issue? can we elaborate?

File: test/ordered_set_test.dart [hunk 2]
```diff
@@ -320,14 +320,24 @@ void main() {
...
         orderedSet.addAll([1, 3, 4]);
         expect(orderedSet.toList().join(), '134');
         expect(orderedSet.remove(4), true);
         expect(orderedSet.toList().join(), '13');
         expect(orderedSet.add(2), true);
         expect(orderedSet.toList().join(), '123');
       });
+
+      test('correct order after remove', () {
+        final set = OrderedSet<int>();
```
Comment (luanpotter) [inline test/ordered_set_test.dart:333]: should we call it orderedSet to be consistent?
`set` trips github highlighting

File: test/ordered_set_test.dart (hunk 1 above)
Comment (spydon) [reply test/ordered_set_test.dart:323]: Tried again, and it wasn't even needed to reproduce the issue, so I just removed the comment.

File: test/ordered_set_test.dart (hunk 2 above)
Comment (spydon) [reply test/ordered_set_test.dart:333]: Fixed

**Pull request merged**